chain/network/Cargo.toml

```diff
@@ -21,6 +21,7 @@ tracing = "0.1.13"
 strum = { version = "0.20", features = ["derive"] }
 near-rust-allocator-proxy = "0.2.9"
 bytesize = "1.0.1"
+conqueue = "0.4.0"
```

Comment:
How confident are we about this crate? cc @matklad

Reply:
@bowenwang1996 It's a small package just only 200 lines of code. I read the code it looks good to me.

Reply:
Yeah, I'd probably go for https://docs.rs/crossbeam/0.8.0/crossbeam/queue/index.html instead. That's 200 lines of `unsafe` and relaxed atomics, and that's hard to get right. crossbeam channel received much more scrutiny . Additionally, each new crate is a supply chain attack vector -- you can steal/fish crates.io credentials, publish a patch release, and it'll get picked up by a next cargo update. Popular crates also suffer from this problem, but their releases have more eyes on them. But we already have crossbeam as a transitive dependency. 

Crossbeam is mpmc, while this is mpsc, so in theory there might be a perf difference, but for that we need to look at the benchmarks first. 

More generally, I am a bit unnerved by the number of dependencies we already have. There's 3 different versions of rand, for example :)

Reply:
@pmnoxx

Reply:
I looked at that library, it's using a spin lock, which could cause serious performance issues if another thread gets suspended by the operating system while having the lock. I would not recommend to use it. I would be if sacrificing performance, if it means that the locking mechanism by a constant factor of 10 times slower. However, using spin lock can potentially cause delays of up to few hundreds milliseconds.  This can happen often since` EdgeVerifier` can request lock quite often.

I saw `Actix` uses similar approach like `conqueue`: 
https://github.com/actix/actix/blob/master/actix/src/address/queue.rs
It's a lock-free mpsc queue, they copied data from: https://docs.rs/futures/0.1.31/src/futures/unsync/mpsc.rs.html#1-474
This approach doesn't use spin lock, which so this is a preferred way for process communication.

If we don't want to use `conqueue`, then I think we should copy the implementation from https://docs.rs/futures/0.1.31/src/futures/unsync/mpsc.rs.html
That approach has been well tested, and it's a preferred way for multi thread communication, which both `actix` and `futures` use.

We could extend this implement prioritization. Instead of having one queue, we could have a set of `k` queues, which priorities 0 to k -1. We could for example, every 1ms check whenever any of the queues has data, process data from lowest queue, and keep doing that until all queue are empty. That way we could add prioritization on top of Actix/Tokio. If I were to implement my own IO loop, I would also use queues like that as well.

Reply:
Oh dear, unbounded spin-loops are indeed a pretty serious problem. I am surprised, to say the least, to learn that crossbeam seg-queue is susceptible, without this being mentioned in the docs.  Opened https://github.com/crossbeam-rs/crossbeam/issues/675

Yeah, in this case this dependency seems reasonable!